### PR TITLE
Lambda

### PR DESCRIPTION
**Pull request by polmacaoidh**

### CODE REVIEW

File: ais-batch-lambda/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import io.spring.gradle.dependencymanagement.dsl.DependencyManagementExtension
+
+plugins {
+    java
+    id("io.spring.dependency-management").version("1.0.6.RELEASE")
```
Comment (JoelClemence) [inline ais-batch-lambda/build.gradle.kts]: Other plugins use the resolution strategy in the settings.gradle to determine the version should this be the same for this? Should the version be added to ```buildSrc/src/main/kotlin/Versions.kt``` like other versions?

File: ais-batch-lambda/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+configure<DependencyManagementExtension> {
+    imports {
+        mavenBom("com.amazonaws:aws-java-sdk-bom:1.11.228")
+    }
+}
+
+dependencies {
+    testImplementation("junit:junit:4.12")
+    implementation("com.amazonaws:aws-java-sdk-emr")
+    implementation("com.amazonaws:aws-lambda-java-core:1.2.0")
```
Comment (JoelClemence) [inline ais-batch-lambda/build.gradle.kts]: Should this be using a different version? Should this use the versions file?

File: ais-batch-lambda/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    isZip64 = true
+}
+
+repositories {
+    mavenCentral()
+}
+
+configure<DependencyManagementExtension> {
+    imports {
+        mavenBom("com.amazonaws:aws-java-sdk-bom:1.11.228")
```
Comment (JoelClemence) [inline ais-batch-lambda/build.gradle.kts]: Should the version come from the Versions file?

File: ais-batch-lambda/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    }
+}
+
+dependencies {
+    testImplementation("junit:junit:4.12")
```
Comment (JoelClemence) [inline ais-batch-lambda/build.gradle.kts]: Should this come from the versions file? I have the same in my branch so we are now in a race condition!

File: buildSrc/src/main/kotlin/Versions.kt [hunk 5]
```diff
@@ -7,9 +7,13 @@ object Versions {
...
     const val geotrellis = "2.2.0"
+    const val mavenAwsBom = "1.11.228"
+    const val junit = "4.12"
+    const val lambdaJavaCore = "1.2.0"
 
     // Plugins
     const val spotless = "3.23.0"
     const val shadow = "5.0.0"
     const val wartremover = "0.9.6"
+    const val springDependancyManagement = "1.0.6.RELEASE"
```
Comment (JoelClemence) [inline buildSrc/src/main/kotlin/Versions.kt]:
```suggestion
    const val springDependencyManagement = "1.0.6.RELEASE"
```

File: settings.gradle.kts [hunk 6]
```diff
@@ -5,6 +5,7 @@ pluginManagement {
                 "com.diffplug.gradle.spotless" -> useVersion(Versions.spotless)
                 "com.github.johnrengelman.shadow" -> useVersion(Versions.shadow)
                 "cz.augi.gradle.wartremover" -> useVersion(Versions.wartremover)
+                "io.spring.dependency-management" -> useVersion(Versions.springDependancyManagement)
```
Comment (JoelClemence) [inline settings.gradle.kts]:
```suggestion
                "io.spring.dependency-management" -> useVersion(Versions.springDependencyManagement)
```

File: ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    public String handleRequest(final Integer inputInt, final Context context) {
+
+        AmazonElasticMapReduce emr = buildMapReduceClient();
+
+        List<Application>  apps = buildApplications();
+
+        List<StepConfig> buildSteps = buildSteps();
+
+        RunJobFlowRequest request = buildJobFlowRequest(buildSteps, apps);
+
```
Comment (ciaranevans) [inline ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java]: Can all of the above not be final?

File: ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    private static final String CLUSTER_NAME = System.getenv("CLUSTER_NAME");
+    private static final String EMR_VERSION = System.getenv("EMR_VERSION");
+    private static final String INSTANCE_COUNT = System.getenv("INSTANCE_COUNT");
+
+    @Override
+    public String handleRequest(final Integer inputInt, final Context context) {
+
+        AmazonElasticMapReduce emr = buildMapReduceClient();
+
+        List<Application>  apps = buildApplications();
```
Comment (ciaranevans) [inline ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java]: Double space between > and apps

File: ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    private List<Application> buildApplications() {
+
+        return Collections.singletonList(new Application().withName("spark"));
+    }
+
+    private List<StepConfig> buildSteps() {
+
+        HadoopJarStepConfig sparkStepConf = new HadoopJarStepConfig()
```
Comment (ciaranevans) [inline ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java]: Final?

File: ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                .withJar("command-runner.jar")
+                .withArgs("spark-submit",
+                        "--class", JOB_FULLY_QUALIFIED_CLASS_NAME,
+                        JOB_LOCATION,
+                        "-i", INPUT_LOCATION,
+                        "-o", OUTPUT_LOCATION,
+                        "-r", RESOLUTION);
+
+        StepConfig sparkStep = new StepConfig()
```
Comment (ciaranevans) [inline ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java]: Final?

File: src/main/terraform/infra_deployment/modules/emr_lambda/main.tf [hunk 11]
```diff
@@ -24,7 +24,25 @@ resource "aws_lambda_function" "emr_lambda" {
   role          = "${aws_iam_role.iam_for_emr_lambda.arn}"
   handler       = "uk.gov.ukho.aisbatchlambda.AisBatchLambdaHandler"
   runtime       = "java8"
+  timeout       = 303
```
Comment (ciaranevans) [inline src/main/terraform/infra_deployment/modules/emr_lambda/main.tf]: Seems quite a random figure?

File: ais-batch-lambda/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,33 @@
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import io.spring.gradle.dependencymanagement.dsl.DependencyManagementExtension
+
+plugins {
+    java
+    id("io.spring.dependency-management")
```
Comment (ciaranevans) [inline ais-batch-lambda/build.gradle.kts]: Might be me being dumb, where are we using spring anywhere?

**Comment by JoelClemence** [comment]
Should we add Checkstyle / Findbugs to this given it is a Java module? I know there are issues to add this to the scala job too but we can limit to the lambda for now. But it would pick up the issues that Ciaran has picked up.

**Comment by eganjs** [comment]
> Should we add Checkstyle / Findbugs to this given it is a Java module? I know there are issues to add this to the scala job too but we can limit to the lambda for now. But it would pick up the issues that Ciaran has picked up.

Great idea

File: ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+import com.amazonaws.services.elasticmapreduce.model.JobFlowInstancesConfig;
+import com.amazonaws.services.elasticmapreduce.model.RunJobFlowRequest;
+import com.amazonaws.services.elasticmapreduce.model.StepConfig;
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+
+import java.util.Collections;
+import java.util.List;
+
+public class AisBatchLambdaHandler implements RequestHandler<Integer, String> {
```
Comment (JoelClemence) [inline ais-batch-lambda/src/main/java/uk/gov/ukho/aisbatchlambda/AisBatchLambdaHandler.java]: make final? 🤣

File: build.gradle.kts [hunk 14]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (JoelClemence) [inline build.gradle.kts]: Needed?

**Pull request merged**